R/brace_linter.R

```diff
@@ -29,9 +29,15 @@ brace_linter <- function(allow_single_line = FALSE) {
       (@line1 = following-sibling::expr/OP-LEFT-BRACE/@line1)
     )",
     # allow `(`, `,` and `%>%` on preceding line
+    #
+    # note that '{' is not supported in RHS call of base-R's native pipe (`|>`),
+    # so this doesn't need to be supported
     "not(
       @line1 = parent::expr/preceding-sibling::*[1][
-        self::OP-LEFT-PAREN or self::OP-COMMA or (self::SPECIAL and text() = '%>%')
+        self::OP-LEFT-PAREN or
+        self::OP-COMMA or
```

Comment:
this doesn't quite feel right to me but I cant think of a counterexample at the moment 🤭

would you mind trying to run .dev/compare_branches on a few packages to see what turns up?

my thinking is we _do_ still want to throw a lint if the node before the comment doesn't match these existing criteria. so the XPath (untested) should be something like preceding-sibling::*[not(self::COMMENT)][1][...] where ... is the existing test.

Reply:
> I cant think of a counterexample at the moment 

Is this a good counterexample? This should produce a lint, correct?

``` r
lintr::lint(
  "# bla
  local(
    {
      x <- 2
    }
  )",
  lintr::brace_linter()
)
```

<sup>Created on 2022-07-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
that also doesn't produce a lint on `HEAD`, and is one of the carve-outs we have regarding the preceding line:

https://github.com/r-lib/lintr/blob/75d4d3116107f33f258ff6c0eeb8a65464888fa1/R/brace_linter.R#L31

Reply:
I ran .dev/compare_branches.R on a sample of 100 packages, found an example that lints at HEAD but not in this branch:

https://github.com/cran/gets/blob/beb581d92e3a0d448ef1a69932c98036fcb4baf9/R/gets-base-source.R#L262-L264

The comment does not excuse the incorrect use of `{` where it should be on the previous line with `function ()`

Reply:
Awesome! Thanks for finding a counterexample. I have added a (currently failing) test for this.

> preceding-sibling::*[not(self::COMMENT)][1][...] 

I tried this, but the test for #1434 fails. I will see how it can be modified to make sure all tests pass.

Reply:
No matter the xpath I try, I seem to get only 2 out 3 tests to pass. So I'm stuck here.

Reply:
Subtle one! we were using a fixed test of `@line1 = @prev::line2 + 1`, but where there's an intervening comment (or comments), `+1` won't be right.

This condition is to lint on cases like `foo(x, { y })` where `,` and `{` are on the same line, so we don't actually care about `+1` -- we only care that `@line1 != @prev::line2`.